dashboard/app/models/user.rb

```diff
@@ -876,14 +877,92 @@ def hidden_stage?(script_level)
     if !script_sections.empty?
       # if we have one or more sections matching this script id, we consider a stage hidden if all of those sections
       # hides the stage
-      script_sections.all? {|s| script_level.stage_hidden_for_section?(s.id)}
+      script_sections.all? {|s| script_level.hidden_for_section?(s.id)}
     else
       # if we have no sections matching this script id, we consider a stage hidden if any of the sections we're in
       # hide it
-      sections.any? {|s| script_level.stage_hidden_for_section?(s.id)}
+      sections.any? {|s| script_level.hidden_for_section?(s.id)}
     end
   end
 
+  # This method will extract a list of hidden ids by section. The type of ids depends
+  # on the input. If hidden_stages is true, id is expected to be a script id and
+  # we look for stages that are hidden. If hidden_stages is false, id is expected
+  # to be a course_id, and we look for hidden scripts.
+  # @param {string} assign_id - Course/script id we're asking for hidden elements of
+  # @param {boolean} [hidden_stages] - Whether we're looking for hidden stages or scripts
+  # @return {Hash<string,number[]>|number[]}
+  #   For teachers, this will be a hash mapping from section id to a list of hidden
+  #   ids for that section.
+  #   For students this will just be a lilst of ids that are hidden for them.
+  def get_hidden_ids(assign_id, hidden_stages = true)
+    # If we're a teacher, we want to go through each of our sections and return
+    # a mapping from section id to hidden stages/scripts in that section
+    return get_teacher_hidden_ids(hidden_stages) if teacher?
+
+    # if we're a student, we want to look through each of the sections in which
+    # we're a member, and use those to figure out which stages should be hidden
+    # for us
+    sections = sections_as_student
+    return [] if sections.empty?
+
+    assigned_sections = sections.select do |section|
+      hidden_stages && section.script_id == assign_id || !hidden_stages && section.course_id == assign_id
+    end
+
+    if assigned_sections.empty?
+      # if we have no sections matching this assignment, we consider a stage/script
+      # hidden if any of our sections hides it
+      return (hidden_stages ? hidden_stage_ids(sections) : hidden_script_ids(sections)).uniq
+    else
+      # if we do have sections matching this assignment, we consider a stage/script
+      # hidden only if it is hidden in every one of the sections the student belongs
+      # to that match this assignment
+      all_ids = hidden_stages ? hidden_stage_ids(assigned_sections) : hidden_script_ids(assigned_sections)
+
+      counts = all_ids.each_with_object(Hash.new(0)) {|id, hash| hash[id] += 1}
+      return counts.select {|_, val| val == assigned_sections.length}.keys
+    end
+  end
```

Comment:
I think this might be even more readable if it's a single call to one or another private method based on user type, rather than one helper and the other inline:

Something like
```ruby
teacher? ? 
  get_teacher_hidden_ids(hidden_stages) : 
  get_student_hidden_ids(assign_id, hidden_stages)
```